Initial release

beck opened the pull request:
A rewrite of [grunt-static-i18n](https://github.com/beck/grunt-static-i18n) that uses streams and adds support for handlebars.

As this is my first gulp plugin, I would like to share what I learned:

When discussing the switch from Grunt > Gulp, I welcomed the change because I found gulpfiles to be easier to read—there is no confusion between task names and option names and streams present a clear flow of data.

IMO, that clarity is the result of pushing complexity back into the plugins. Steams come with a high learning curve. When tearing down the syntax of a gulp task:

``` js
gulp.src('somefile')
  .pipe(plugin({some: config});
```

We are making a call to the `plugin` module to return a stream transformer, a function that returns a function.  All requiring the use of [through2](https://github.com/rvagg/through2/blob/master/through2.js) - functions returning functions returning functions .... turtles all the way down.  :turtle: :turtle: :turtle: 

Overall, I think gulp is the better build tool—just more difficult to write for.  I relied heavily on tests to give me confidence all is correct, if anything here is confusing let me know.

beck committed:
Add boilerplate

beck committed:
Scaffold plugin

beck committed:
Create a translator.

* Provides interface for the gettext catalogs.
* Provides a transformer, this is the function used when piping streams.
  In this case, untranslated files will be piped in and translated
  copy will be emitted.

beck committed:
Rename translate lib to be an obj

beck committed:
Update stream translator to push instead of emit

beck committed:
Add default language (en) to translator

beck committed:
Write translated files to disk

beck committed:
Translate handlebars

beck committed:
Move file type translators to their own source.

This cleans up the code and fixes a bug.

Bug fixed: non-supported files are not moved into translated source. Now
all files are passed through the translator.

By "registering" translators to certain file extensions and passing all
files through the translator, this removes the need to use promises.

beck committed:
Update docs

imorrison commented on .travis.yml:
```diff
@@ -0,0 +1,6 @@
+sudo: false
+language: node_js
+node_js:
+  - 'iojs'
```
Now you are just getting hipster on us.

blaix replied:
https://twitter.com/naholyr/status/569716245697851392

kahnvex replied:
Nodejs devs be like 

![old](http://i.giphy.com/BgBf6pW9qOgQU.gif)

kahnvex commented on test/fixtures/app/gulpfile.js:
```diff
@@ -0,0 +1,27 @@
+'use strict';
+var gulp = require('gulp');
+var clean = require('gulp-rimraf');
+var statici18n = require('../../../');
```
Whats this?

beck replied:
This is the test fixture importing the plugin via [index.js](https://github.com/yola/gulp-static-i18n/blob/9aba247446b58e14b694a56ce57d9d1735e12a41/index.js).

kahnvex replied:
Ok, I think doing `require('../../../index.js');` might make this a bit more clear.

beck committed:
Minor cleanup of imports and spelling

ahoym commented on README.md:
````diff
@@ -0,0 +1,105 @@
+```js
+'use strict';
+var gulp = require('gulp');
+var clean = require('gulp-rimraf');
````
`gulp-rimraf` is being deprecated in favor of [del](https://github.com/gulpjs/gulp/blob/master/docs/recipes/delete-files-folder.md)

beck replied:
gah, gulp-clear deprecated for gulp-rimraf deprecated for del, sheesh

ahoym commented on lib/translator.js:
```diff
@@ -0,0 +1,141 @@
+
+  var localeDir = this.options.localeDir;
+  var isLocale = function(file) {
+    var filePath = localeDir + '/' + file;
```
`path.join(localeDir, file)` ?

**Edit**
Line 49 too

ahoym commented on lib/translator.js:
```diff
@@ -0,0 +1,141 @@
+
+Translator.prototype.getTranslator = function(file) {
+  var ext = path.extname(file.path);
+  return this.translators[ext] || lodash.identity;
```
Missing argument for `lodash.identity` ?

beck replied:
So we're getting a translator here based on the file extension, returning a function.

Currently there are two: a js translator and an hbs translator (and once this lands, I will be making a third for json).

If there is no translator (say a .txt file is passed) then the translator needs to return the passed-in copy, no changes, aka an identity function.

imorrison replied:
> So we're getting a translator here based on the file extension, returning a function.

Should that assumptions be made? You could also have the contents piped into different translators?

imorrison replied:
Yea, I think you are abstracting to much for the user of this plugin. I think it would be nice to have the ability to pipe whatever translator you wanted. Your translate functions have a fairly simple interface.

beck replied:
> Should that assumption be made?

I assume you mean file extensions?  I think so.  I tried to build this plugin with minimal configuration and produce an initial release based on solely convention.

> have the contents piped into different translators?

Yes.  Check out [index.js](https://github.com/yola/gulp-static-i18n/blob/doug/init/index.js#L34) where I register translators to file extensions.

> I think you are abstracting to much

My [first iteration has no such abstraction](https://github.com/yola/gulp-static-i18n/commit/b918c68a8ad01f0473395f2f037f71e1997cb47f#diff-168726dbe96b3ce427e7fedce31bb0bcL5) but I was using q and promises to ensure each file type was finished translating before complete. Extracting the file type specific code to their own domain made it easy to factor away any need for promises.

> translate functions have a fairly simple interface

True the [js translator](https://github.com/yola/gulp-static-i18n/blob/ab697b214f6866681b2cc1c75527460ee5033878/lib/translators/javascript.js) and the [handlebars translator](https://github.com/yola/gulp-static-i18n/blob/doug/init/lib/translators/handlebars.js) are simple but also extremely limited.  Each may become increasing complex as we want to support such things as:
- String addition inside the gettext/trans call
- Whitespace reduction
- Pluralization
- Replace [greedy regex](https://github.com/yola/gulp-static-i18n/blob/doug/init/test/test-transjs.js#L27) with a proper syntax parser

These changes can now happen in in their own files without touching the core plugin code. It also makes it super simple to add support for new file types, which I'll be doing very soon with JSON.

I'm very much open to suggestion / cirque here.

kahnvex commented:
Tests are broken for me, had to install jshint.

beck committed:
Add jshint to deps

beck committed:
Test using "expect" not "should".

@beck: I started using "should" with my tests.  I like it.
@blaix: I feel the exact opposite
@blaix: it "should
@blaix: it "should
@blaix: it "should
@blaix: boring noise
@beck: you like expect better?
@blaix: yes
@blaix: it follows much better OO principles
@blaix: it doesn't pollute every object
@blaix: expect() decorates objects
@blaix: much better
@beck: your mom is better.

beck committed:
Use del not gulp-rimraf

beck committed:
Use path module not string addition

beck committed:
Put the default lang it it's own folder, removing special treatment.

When implementing the plugin, it was exceedingly difficult to treat each
language as it's own build if the default language was not in it's own folder.

With this change, it is the responsibility of the app to flatten the default
language.

euwest commented on lib/translated-file.js:
```diff
@@ -0,0 +1,34 @@
+  if(lang === defaultLang) {
+    return '';
+  }
+  var dir = lang.toLowerCase().replace(/_/g, '-');
```
isn't this going to result in urls like `templates.yola.com/pt-br` and `yola.com/cn-zh/support`?

beck replied:
yep yep

beck committed:
Prepare v0.0.1

beck committed:
Add lib dir to package files

beck committed:
Leave untouched files without an associated translator.

imorrison commented:
:+1:
